File: scripts/retro.py
#!/usr/bin/env python3
"""retro.py — v0.6 retrospective ceremony template + events.log analysis.

Fires after `/harness:work F-N --complete` succeeds. Analyzes events.log
to auto-populate machine-extractable sections, leaves Reviewer Reflection
+ Copy Polish for orchestrator (reviewer → tech-writer prose chain).

Usage:
    python3 scripts/retro.py --harness-dir .harness --feature F-N
"""

from __future__ import annotations

import argparse
import json
import sys
from datetime import datetime, timezone
from pathlib import Path


def _read_events(harness_dir: Path) -> list[dict]:
    log_path = harness_dir / "events.log"
    if not log_path.is_file():
        return []
    events = []
    for line in log_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            events.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return events


def analyze(events: list[dict], feature_id: str) -> dict:
    """Extract machine-readable retro data from events.

    Event key contract mirrors scripts/work.py (canonical emitter):
      - field name is `"feature"` (not "feature_id")
      - completion type is `"feature_done"` (not "feature_completed")
    """
    relevant = [e for e in events if e.get("feature") == feature_id]
    gate_events = [e for e in relevant if e.get("type") == "gate_recorded"]
    first_gate_fail: dict | None = None
    for e in gate_events:
        if e.get("result") == "fail":
            first_gate_fail = e
            break
    completed = any(e.get("type") == "feature_done" for e in relevant)
    kickoff_opened = any(e.get("type") == "kickoff_started" for e in relevant)
    design_review_opened = any(
        e.get("type") == "design_review_opened" for e in relevant
    )
    questions_opened = sum(
        1 for e in relevant if e.get("type") == "question_opened"
    )
    questions_answered = sum(
        1 for e in relevant if e.get("type") == "question_answered"
    )
    return {
        "completed": completed,
        "first_gate_fail": first_gate_fail,
        "kickoff_opened": kickoff_opened,
        "design_review_opened": design_review_opened,
        "questions_opened": questions_opened,
        "questions_answered": questions_answered,
        "gate_events_total": len(gate_events),
        "all_events_total": len(relevant),
    }


def _template(feature_id: str, analysis: dict, timestamp: str) -> str:
    fgf = analysis["first_gate_fail"]
    if fgf:
        fgf_line = (
            f"- {fgf.get('gate', '?')} failed at {fgf.get('ts', '?')}"
            f"  (reason: {fgf.get('note') or fgf.get('reason') or '?'})"
        )
    else:
        fgf_line = "- 없음 (전 gate 최초에 pass)"

    lines: list[str] = []
    lines.append(f"# Retrospective — {feature_id}")
    lines.append("")
    lines.append(f"> 자동 생성 — {timestamp}")
    lines.append(">")
    lines.append(
        "> `scripts/retro.py` 가 events.log 를 분석해 머신 섹션을 채우고, "
        "orchestrator 가 reviewer → tech-writer 순차로 Reviewer Reflection · "
        "Copy Polish 섹션을 완성한다."
    )
    lines.append("")

    # What Shipped
    lines.append("## What Shipped")
    lines.append("")
    if analysis["completed"]:
        lines.append(f"- {feature_id} — complete 전이 감지.")
    else:
        lines.append(f"- {feature_id} — complete 이벤트 미감지. 수동 확인 필요.")
    lines.append("")

    # First Gate to Fail
    lines.append("## First Gate to Fail")
    lines.append("")
    lines.append(fgf_line)
    lines.append("")

    # Ceremonies
    lines.append("## Ceremonies")
    lines.append("")
    lines.append(f"- Kickoff opened: {'✅' if analysis['kickoff_opened'] else '❌'}")
    lines.append(
        f"- Design Review opened: {'✅' if analysis['design_review_opened'] else '❌ (해당 피처에 미실행)'}"
    )
    lines.append(
        f"- Questions opened: {analysis['questions_opened']} · answered: {analysis['questions_answered']}"
    )
    lines.append("")

    # Risks Materialized vs plan.md — LLM 섹션
    lines.append("## Risks Materialized vs plan.md")
    lines.append("")
    lines.append(
        "<!-- orchestrator via reviewer: plan.md Risks 와 실제 구현 중 발생 사건 diff -->"
    )
    lines.append("")
    lines.append("_(pending)_")
    lines.append("")

    # Decisions Revised — LLM 섹션
    lines.append("## Decisions Revised")
    lines.append("")
    lines.append(
        "<!-- orchestrator via reviewer: 구현 중 새 ADR 또는 status 전이 (superseded 등) -->"
    )
    lines.append("")
    lines.append("_(pending)_")
    lines.append("")

    # Kickoff predictions — LLM 섹션
    lines.append("## Kickoff Predictions That Were Right / Wrong")
    lines.append("")
    lines.append(
        "<!-- orchestrator via reviewer: kickoff/F-N.md 의 bullets vs 실제 결과 -->"
    )
    lines.append("")
    lines.append("_(pending)_")
    lines.append("")

    # Reviewer Reflection
    lines.append("## Reviewer Reflection")
    lines.append("")
    lines.append(
        "<!-- orchestrator invokes @harness:reviewer to produce draft prose."
        " reviewer 는 read-only (CQS) — draft 텍스트만 반환. orchestrator 가 이 섹션에 write. -->"
    )
    lines.append("")
    lines.append("_(pending)_")
    lines.append("")

    # Copy Polish
    lines.append("## Copy Polish")
    lines.append("")
    lines.append(
        "<!-- orchestrator invokes @harness:tech-writer to polish the Reviewer Reflection."
        " tech-writer 가 Write/Edit 으로 직접 이 섹션을 다듬음. -->"
    )
    lines.append("")
    lines.append("_(pending)_")
    lines.append("")

    return "\n".join(lines).rstrip() + "\n"


def _append_event(harness_dir: Path, event: dict) -> None:
    log_path = harness_dir / "events.log"
    log_path.parent.mkdir(parents=True, exist_ok=True)
    with log_path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(event, ensure_ascii=False) + "\n")


def generate_retro(
    harness_dir: Path,
    *,
    feature_id: str,
    timestamp: str | None = None,
) -> Path:
    if timestamp is None:
        timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    events = _read_events(harness_dir)
    analysis = analyze(events, feature_id)
    retro_dir = harness_dir / "_workspace" / "retro"
    retro_dir.mkdir(parents=True, exist_ok=True)
    path = retro_dir / f"{feature_id}.md"
    path.write_text(_template(feature_id, analysis, timestamp), encoding="utf-8")
    _append_event(
        harness_dir,
        {
            "ts": timestamp,
            "type": "feature_retro_written",
            "feature": feature_id,
            "analysis_summary": {
                "completed": analysis["completed"],
                "first_gate_fail": (
                    analysis["first_gate_fail"].get("gate")
                    if analysis["first_gate_fail"]
                    else None
                ),
                "questions_opened": analysis["questions_opened"],
            },
            "path": str(path.relative_to(harness_dir)),
        },
    )
    return path


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Generate retro template + events.log analysis")
    parser.add_argument("--harness-dir", type=Path, default=Path(".harness"))
    parser.add_argument("--feature", required=True)
    parser.add_argument("--timestamp", default=None)
    args = parser.parse_args(argv)
    path = generate_retro(
        args.harness_dir, feature_id=args.feature, timestamp=args.timestamp
    )
    print(str(path))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/ceremonies/retro.ts
/**
 * Retrospective ceremony template generator (F-097 port of
 * `scripts/ceremonies/retro.py`).
 *
 * Fires after `/harness:work F-N --complete` succeeds. Reads
 * `events.log`, fills the machine-extractable sections (What
 * Shipped, First Gate to Fail, Ceremonies summary), and leaves the
 * LLM-driven sections as `_(pending)_` for orchestrator's reviewer →
 * tech-writer chain.
 *
 * Idempotency contract (v0.8.7): existing retro files are not
 * overwritten unless `force=true`.
 *
 * @module ceremonies/retro
 */

import {appendFileSync, mkdirSync, readFileSync, statSync, writeFileSync} from 'node:fs';
import {dirname, join, relative} from 'node:path';

/** One row from events.log after parsing. */
type EventRecord = Record<string, unknown>;

/** Machine-extractable retro analysis. */
export interface RetroAnalysis {
  completed: boolean;
  first_gate_fail: EventRecord | null;
  kickoff_opened: boolean;
  design_review_opened: boolean;
  questions_opened: number;
  questions_answered: number;
  gate_events_total: number;
  all_events_total: number;
  archived: boolean;
  archived_event: EventRecord | null;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function readEvents(harnessDir: string): EventRecord[] {
  const logPath = join(harnessDir, 'events.log');
  let raw: string;
  try {
    raw = readFileSync(logPath, 'utf-8');
  } catch {
    return [];
  }
  const out: EventRecord[] = [];
  for (const line of raw.split('\n')) {
    const trimmed = line.trim();
    if (trimmed.length === 0) {
      continue;
    }
    try {
      const parsed: unknown = JSON.parse(trimmed);
      if (isPlainObject(parsed)) {
        out.push(parsed);
      }
    } catch {
      continue;
    }
  }
  return out;
}

/**
 * Extracts machine-readable retro data from a list of events for a
 * single feature.
 *
 * Event-key contract (matches `scripts/work.py`):
 *   - feature id key is `feature` (not `feature_id`)
 *   - completion type is `feature_done` (not `feature_completed`)
 *   - archive type is `feature_archived`
 */
export function analyze(events: ReadonlyArray<EventRecord>, featureId: string): RetroAnalysis {
  const relevant = events.filter((e) => e['feature'] === featureId);
  const gateEvents = relevant.filter((e) => e['type'] === 'gate_recorded');
  let firstGateFail: EventRecord | null = null;
  for (const e of gateEvents) {
    if (e['result'] === 'fail') {
      firstGateFail = e;
      break;
    }
  }
  const completed = relevant.some((e) => e['type'] === 'feature_done');
  const kickoffOpened = relevant.some((e) => e['type'] === 'kickoff_started');
  const designReviewOpened = relevant.some((e) => e['type'] === 'design_review_opened');
  const questionsOpened = relevant.filter((e) => e['type'] === 'question_opened').length;
  const questionsAnswered = relevant.filter((e) => e['type'] === 'question_answered').length;
  let archivedEvent: EventRecord | null = null;
  for (const e of relevant) {
    if (e['type'] === 'feature_archived') {
      archivedEvent = e;
    }
  }
  return {
    completed,
    first_gate_fail: firstGateFail,
    kickoff_opened: kickoffOpened,
    design_review_opened: designReviewOpened,
    questions_opened: questionsOpened,
    questions_answered: questionsAnswered,
    gate_events_total: gateEvents.length,
    all_events_total: relevant.length,
    archived: archivedEvent !== null,
    archived_event: archivedEvent,
  };
}

function rstrip(s: string): string {
  return s.replace(/\s+$/, '');
}

function nowIso(): string {
  const d = new Date();
  const yyyy = d.getUTCFullYear().toString().padStart(4, '0');
  const mm = (d.getUTCMonth() + 1).toString().padStart(2, '0');
  const dd = d.getUTCDate().toString().padStart(2, '0');
  const hh = d.getUTCHours().toString().padStart(2, '0');
  const mi = d.getUTCMinutes().toString().padStart(2, '0');
  const ss = d.getUTCSeconds().toString().padStart(2, '0');
  return `${yyyy}-${mm}-${dd}T${hh}:${mi}:${ss}Z`;
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function pythonStyleJsonStringify(value: unknown): string {
  if (value === null) {
    return 'null';
  }
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }
  if (typeof value === 'number') {
    return JSON.stringify(value);
  }
  if (typeof value === 'string') {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((v) => pythonStyleJsonStringify(v)).join(', ')}]`;
  }
  if (typeof value === 'object') {
    const pairs = Object.entries(value as Record<string, unknown>).map(
      ([k, v]) => `${JSON.stringify(k)}: ${pythonStyleJsonStringify(v)}`,
    );
    return `{${pairs.join(', ')}}`;
  }
  throw new TypeError(`retro: unsupported value type ${typeof value}.`);
}

function appendEvent(harnessDir: string, event: Record<string, unknown>): void {
  const logPath = join(harnessDir, 'events.log');
  mkdirSync(dirname(logPath), {recursive: true});
  appendFileSync(logPath, `${pythonStyleJsonStringify(event)}\n`, 'utf-8');
}

function template(
  featureId: string,
  analysis: RetroAnalysis,
  timestamp: string,
  mode: 'product' | 'prototype',
): string {
  const isPrototype = mode === 'prototype';
  const fgf = analysis.first_gate_fail;
  const fgfLine =
    fgf !== null
      ? `- ${(fgf['gate'] as string | undefined) ?? '?'} failed at ${(fgf['ts'] as string | undefined) ?? '?'}` +
        `  (reason: ${(fgf['note'] as string | undefined) || (fgf['reason'] as string | undefined) || '?'})`
      : '- 없음 (전 gate 최초에 pass)';

  const intro = isPrototype
    ? '프로토타입 모드 — 머신 섹션만 자동 채움. LLM 반성 섹션은 생략.'
    : '`scripts/retro.py` 가 events.log 를 분석해 머신 섹션을 채우고, ' +
      'orchestrator 가 reviewer → tech-writer 순차로 Reviewer Reflection · ' +
      'Copy Polish 섹션을 완성한다.';

  const lines: string[] = [];
  lines.push(`# Retrospective — ${featureId}`);
  lines.push('');
  lines.push(`> 자동 생성 — ${timestamp} · mode: \`${mode}\``);
  lines.push('>');
  lines.push(`> ${intro}`);
  lines.push('');

  // What Shipped
  lines.push('## What Shipped');
  lines.push('');
  if (analysis.completed) {
    lines.push(`- ${featureId} — complete 전이 감지.`);
  } else {
    lines.push(`- ${featureId} — complete 이벤트 미감지. 수동 확인 필요.`);
  }
  lines.push('');

  // Superseded By — v0.10
  if (analysis.archived) {
    const ev = analysis.archived_event ?? {};
    const sb = ev['superseded_by'];
    const reason = (ev['reason'] as string | undefined) ?? '(reason 미기록)';
    const ts = (ev['ts'] as string | undefined) ?? '?';
    lines.push('## Superseded By');
    lines.push('');
    if (typeof sb === 'string' && sb.length > 0) {
      lines.push(`- 이 피처는 **${sb}** 로 대체됨 (${ts})`);
    } else {
      lines.push(
        `- 이 피처는 archived 됨 (${ts}) — superseded_by 미지정 (deprecation only · 대체 피처 없음)`,
      );
    }
    lines.push(`- 사유: ${reason}`);
    lines.push('');
    lines.push(
      '<!-- F-028: feature_archived event 의 superseded_by/reason 자동 채움. 수동 추가 컨텍스트는 아래에 자유 기술. -->',
    );
    lines.push('');
  }

  // First Gate to Fail
  lines.push('## First Gate to Fail');
  lines.push('');
  lines.push(fgfLine);
  lines.push('');

  // Ceremonies
  lines.push('## Ceremonies');
  lines.push('');
  lines.push(`- Kickoff opened: ${analysis.kickoff_opened ? '✅' : '❌'}`);
  lines.push(
    `- Design Review opened: ${
      analysis.design_review_opened ? '✅' : '❌ (해당 피처에 미실행)'
    }`,
  );
  lines.push(
    `- Questions opened: ${analysis.questions_opened} · answered: ${analysis.questions_answered}`,
  );
  lines.push('');

  if (isPrototype) {
    return `${rstrip(lines.join('\n'))}\n`;
  }

  // LLM sections — left as pending placeholders.
  for (const heading of [
    'Risks Materialized vs plan.md',
    'Decisions Revised',
    'Kickoff Predictions That Were Right / Wrong',
  ]) {
    lines.push(`## ${heading}`);
    lines.push('');
    lines.push(`<!-- orchestrator via reviewer: ${heading} 섹션 -->`);
    lines.push('');
    lines.push('_(pending)_');
    lines.push('');
  }

  lines.push('## Reviewer Reflection');
  lines.push('');
  lines.push(
    '<!-- orchestrator invokes @harness:reviewer to produce draft prose. ' +
      'reviewer 는 read-only (CQS) — draft 텍스트만 반환. orchestrator 가 이 섹션에 write. -->',
  );
  lines.push('');
  lines.push('_(pending)_');
  lines.push('');

  lines.push('## Copy Polish');
  lines.push('');
  lines.push(
    '<!-- orchestrator invokes @harness:tech-writer to polish the Reviewer Reflection. ' +
      'tech-writer 가 Write/Edit 으로 직접 이 섹션을 다듬음. -->',
  );
  lines.push('');
  lines.push('_(pending)_');
  lines.push('');

  return `${rstrip(lines.join('\n'))}\n`;
}

/** Optional input for {@link generateRetro}. */
export interface GenerateRetroOptions {
  timestamp?: string;
  force?: boolean;
  mode?: 'product' | 'prototype';
}

/**
 * Generates `.harness/_workspace/retro/F-N.md` plus a
 * `feature_retro_written` event. Idempotent — existing retro files
 * are preserved unless `force=true`.
 */
export function generateRetro(
  harnessDir: string,
  featureId: string,
  options: GenerateRetroOptions = {},
): string {
  const timestamp = options.timestamp ?? nowIso();
  const force = options.force ?? false;
  const mode = options.mode ?? 'product';

  const retroDir = join(harnessDir, '_workspace', 'retro');
  mkdirSync(retroDir, {recursive: true});
  const path = join(retroDir, `${featureId}.md`);

  if (isFile(path) && !force) {
    return path;
  }

  const events = readEvents(harnessDir);
  const analysis = analyze(events, featureId);
  writeFileSync(path, template(featureId, analysis, timestamp, mode), 'utf-8');
  appendEvent(harnessDir, {
    ts: timestamp,
    type: 'feature_retro_written',
    feature: featureId,
    mode,
    analysis_summary: {
      completed: analysis.completed,
      first_gate_fail:
        analysis.first_gate_fail !== null ? analysis.first_gate_fail['gate'] : null,
      questions_opened: analysis.questions_opened,
    },
    path: relative(harnessDir, path),
  });
  return path;
}
